### PR TITLE
Graph: Fixes so only users with correct permissions can add annotations

### LINKED ISSUE #29972: Annotations: Viewer can open "Add Annotation" modal (but they cannot create annotation, which is correct)
**What happened**:
Users with the role viewer are allowed and able to create annotations but when they click the save-button they receive the message "Access denied to this dashboard".

annotations
 
**What you expected to happen**:
The expected behaviour is that once the user with viewer role clicks the save-button that the annotation is saved. The alert pop-up should then say: "Annotation added" and have a green background.

**How to reproduce it (as minimally and precisely as possible)**:
* Log in as a user with the role viewer
* open a dashboard with a graph panel
* click the graph panel and select 'add annotation'
* add a description
* click save-button

**Anything else we need to know?**:
I found this issue: https://github.com/grafana/grafana/issues/9991
This issue is labeled as a a feature request, however, it does not mention that a viewer can already add an annotation, but cannot save it.

**Environment**:
- Grafana version: 7.3.6
- Data source type & version: data source used -- grafana -- with query randomwalk to create the graph panel
- OS Grafana is installed on: ms windows os version 10
- User OS & Browser: ms windows os version 10 - firefox
- Grafana plugins: default plugins
- Others: none

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
When fixing https://github.com/grafana/grafana/pull/29990 I forgot to add cases for when users don't have permissions to add annotations. This PR extracts the logic for when a user can add an annotation into the `canAddAnnotations` function and uses it accordingly where needed.

This is probably something that needs to migrate to GraphNG moving forward.

Enjoy the review 🎉 

**Which issue(s) this PR fixes**:
Fixes #29972

**Special notes for your reviewer**:
- test adding annotation by clicking in Graph (viewers shouldn't see Add annotations menu item)
- test adding annotation by selecting an area in the graph while holding down `cmd|ctrl` key
- test adding annotation by clicking in Graph while while holding down `cmd|ctrl` key
- test it with a user that has edit permissions
- test it with a user that has edit permissions and make the dashboard not editable => Settings => General => Editable toggle
image
- test it with a user that has viewer permissions

### CODE REVIEW

File: public/app/plugins/panel/graph/graph.ts [hunk 1]
```diff
@@ -925,6 +925,10 @@ class GraphElement {
       return formattedValueToString(formatter(val, axis.tickDecimals, axis.scaledDecimals));
     };
   }
+
+  canAddAnnotations() {
+    return (this.dashboard.meta.canEdit || this.dashboard.meta.canMakeEditable) && this.dashboard.editable;
```
Comment (torkelo) [inline public/app/plugins/panel/graph/graph.ts:930]: this.dashboard.editable should be here, then you cannot add annotations to dashboards that are not editable (which you should be able to do)

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts:930]: Sorry I don't understand, but `this.dashboard.editable` is part of the logic?

Comment (torkelo) [reply public/app/plugins/panel/graph/graph.ts:930]: yes, think (this.dashboard.meta.canEdit || this.dashboard.meta.canMakeEditable)  should cover it

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts:930]: It doesn't, unfortunately. It's very confusing with all these props. But when `editable = false` then an editor could still add annotations. See #29990.

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts:930]: I'll do another test run to confirm

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts:930]: I did another test run without `this.dashboard.editable` in the statement and it fails for non editable dashboards. For non editable dashboards someone with edit permissions would still be able to add annotations so `this.dashboard.editable` still needs to be in the statement.

Comment (torkelo) [reply public/app/plugins/panel/graph/graph.ts:930]: But that seems to be the current and correct behavior (that you can still add annotations to noneditable dashboards):
https://play.grafana.org/d/Z9NCtxBMz/test-editable?orgId=1

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts:930]: Probably because #29990 hasn't been released yet.

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts:930]: Hmm, now I'm a bit confused. Then this issue is not a correct then https://github.com/grafana/grafana/issues/28738

Comment (torkelo) [reply public/app/plugins/panel/graph/graph.ts:930]: yea, think the bug in https://github.com/grafana/grafana/issues/28738 was not that you cannot add an annotation but that you cannot then delete or edit it

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts:930]: Ah I see, I'll try to fix this in this PR as well. Thank you for clarifing this.

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts]: I've moved the `canAddAnnotations` function to the `DashboardModel` instead and added a check for this in annotation_tooltip as well.

Comment (hugohaggmark) [reply public/app/plugins/panel/graph/graph.ts]: This means that 7.4.0-beta1 will have the intermittent behavior from #29990 until this PR gets merged. 
How do we do with back port labels and betas? Should I also add a back port label to this PR @torkelo?

Comment (torkelo) [reply public/app/plugins/panel/graph/graph.ts]: yes, that is needed

**Review by torkelo — APPROVED** [review]
look good

**Pull request merged**